vendor: respect repository field in restore and update

cryptix opened the pull request:
Somehow the `repository` field from the manifest isn't used.

With this I can replace packages with forks.

cryptix committed:
respect repository field in restore and update

davecheney commented:
I deeply regret adding restore to gb vendor, thanks for fixing this, I’ll take a look soon. 

> On 11 Jan 2018, at 05:45, Henry <notifications@github.com> wrote:
> 
> Somehow the repository field from the manifest isn't used.
> 
> With this I can replace packages with forks.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/constabulary/gb/pull/730
> 
> Commit Summary
> 
> respect repository field in restore and update
> File Changes
> 
> M cmd/gb-vendor/restore.go (8)
> M cmd/gb-vendor/update.go (2)
> Patch Links:
> 
> https://github.com/constabulary/gb/pull/730.patch
> https://github.com/constabulary/gb/pull/730.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cryptix commented:
> I deeply regret adding restore to gb vendor, ...

oh why is that?

I might not be best style but I tend to keep small patches inside my tree until I move them to PRs/external forks. It reduces the overhead of dealing with a lot more repos for me and the restore command is pretty helpful to get back the upstream state.

davecheney commented:
It has been the biggest source of bugs, especially when people became so reliant on it that parallel downloading of artifacts had to be implemented. 


> On 11 Jan 2018, at 19:02, Henry <notifications@github.com> wrote:
> 
> I deeply regret adding restore to gb vendor.
> 
> oh why is that?
> 
> I might not be best style but I tend to keep small patches inside my tree until I move them to PRs/external forks. It reduces the overhead of dealing with a lot more repos for me and the restore command is pretty helpful to get back the upstream state.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cryptix commented:
~~I just noticed an issue with my approach. Consider these two entries in `manifest`, they retrieve two packages from the same repository.~~

~~Since I just replaced `importPath` with `repository`, it now clones the root of the repositories into the two `improtpath` locations, leading to a broken vendor state.~~

Disregard the previous comment. Somehow my `vendor/manifest` didn't have `"path": "..."` entries to build the correct copy.

cryptix committed:
vendor: special case go.googlesource.com repo deduction

I broke the vanity-detection fallback with my previous commit.
This is just a workaround until there is a cleaner solution.

cryptix commented:
Testing this some more, I noticed that I broke vanity imports.

I pushed a really dirty hack in 0153a33 to make `golang.org/x/...` work again but it's not good since it only works for `go.googlesource.com/...`.
